File: frontend/src/utils/foodTiming.js
/**
 * When a dose is taken relative to food.
 *
 * The vocabulary is the server's (entity/FoodTiming). It lives in its own column rather than
 * inside the free-text instructions field, because that field is general — "take with plenty of
 * water", "crush before giving" — and turning it into a four-value dropdown would have removed
 * the ability to record any of that.
 */
export const FOOD_TIMING_OPTIONS = [
  ['BEFORE_FOOD', 'Before food'],
  ['AFTER_FOOD', 'After food'],
  ['WITH_FOOD', 'With food'],
  ['NOT_SPECIFIED', 'Not specified'],
];

const LABELS = Object.fromEntries(FOOD_TIMING_OPTIONS);

/**
 * Medicine types and routes that bypass the gut, for which "before or after food" states
 * nothing. Both lists are the server's vocabulary (dto/AddIpdPrescriptionRequest), not just
 * the subset the IPD form currently offers, so an order entered through any future control
 * is judged by the same rule.
 */
const PARENTERAL_TYPES = new Set(['INJECTION', 'IV_FLUID']);
const PARENTERAL_ROUTES = new Set(['IV', 'IM', 'SUBCUTANEOUS']);

/**
 * Whether food timing is a meaningful question for this order.
 *
 * An injected dose never meets a meal, so recording "after food" against one is not a
 * harmless extra: it reaches the nurse's medication chart and reads as an instruction.
 * Type and route are checked independently because they can disagree — picking an
 * injection from the catalogue sets the type and leaves the route on its ORAL default.
 *
 * Unknown, absent and blank values are treated as applicable: this hides a question that
 * cannot apply, and must never hide one that might.
 */
export const isFoodTimingApplicable = (type, route) => {
  const norm = (v) => (v == null ? '' : String(v).trim().toUpperCase());
  return !PARENTERAL_TYPES.has(norm(type)) && !PARENTERAL_ROUTES.has(norm(route));
};

/**
 * How to show a stored value.
 *
 * Anything unrecognised is shown as it was written. Orders predate this field, and an older
 * client could still send something else; a clinical record shows what it holds rather than
 * hiding a value it does not recognise.
 */
export const describeFoodTiming = (value) => {
  if (!value) return null;
  return LABELS[value] ?? String(value);
};

export default describeFoodTiming;

File: backend/src/main/java/com/hms/entity/FoodTiming.java
package com.hms.entity;

/**
 * When a dose is taken relative to food.
 *
 * <p>Deliberately a small closed vocabulary held as strings on {@link Prescription}, not a JPA
 * enum. Orders written before this existed carry nothing, and orders written by an older client
 * could carry anything; a strict enum column would fail to deserialise those rows rather than
 * showing them, which is the opposite of what a clinical record should do. Validation happens on
 * the way in, and reading is permissive.
 */
public final class FoodTiming {

    public static final String BEFORE_FOOD = "BEFORE_FOOD";
    public static final String AFTER_FOOD = "AFTER_FOOD";
    public static final String WITH_FOOD = "WITH_FOOD";
    public static final String NOT_SPECIFIED = "NOT_SPECIFIED";

    private static final java.util.Set<String> VALUES =
            java.util.Set.of(BEFORE_FOOD, AFTER_FOOD, WITH_FOOD, NOT_SPECIFIED);

    /**
     * Accept a recognised value, reject anything else, and treat blank as "not stated".
     *
     * <p>Blank returns null rather than NOT_SPECIFIED: an order where nobody answered the question
     * and one where somebody answered "it does not matter" are different facts, and only the
     * second is a clinical statement.
     */
    public static String normalize(String raw) {
        if (raw == null || raw.isBlank()) return null;
        String value = raw.trim().toUpperCase();
        if (!VALUES.contains(value)) {
            throw new IllegalArgumentException(
                    "Food timing must be one of BEFORE_FOOD, AFTER_FOOD, WITH_FOOD or NOT_SPECIFIED");
        }
        return value;
    }

    public static boolean isValid(String value) {
        return value != null && VALUES.contains(value);
    }

    private FoodTiming() {
    }
}
